File: lab2/src/commands/CommandManager.ts
import { Command } from './Command';
import { Editor } from '../editor/Editor';

export class CommandManager {
    private undoStack: Command[] = [];
    private redoStack: Command[] = [];
    private editor: Editor;

    constructor(editor: Editor) {
        this.editor = editor;
    }

    clear(): void {
        this.undoStack = [];
        this.redoStack = [];
    }

    canUndo(): boolean { return this.undoStack.length > 0; }
    canRedo(): boolean { return this.redoStack.length > 0; }

    
    execute(cmd: Command): void {
        try {
            cmd.execute(this.editor);
            this.undoStack.push(cmd);
            this.redoStack = [];
        } catch (err) {
            console.error(err);
        }
    }

    undo(): Command | null {
        const cmd = this.undoStack.pop();
        if (!cmd) return null;
        cmd.undo(this.editor);
        this.redoStack.push(cmd);
        return cmd;
    }

    redo(): Command | null {
        const cmd = this.redoStack.pop();
        if (!cmd) return null;
        cmd.execute(this.editor);
        this.undoStack.push(cmd);
        return cmd;
    }
}

File: member_djy/23302010083_邓嘉宜/src/main/java/com/editor/core/command/CommandManager.java
package com.editor.core.command;

import java.util.Stack;

/**
 * 命令管理器
 * 维护全局命令的undo/redo栈
 * 注意：这里管理的是全局命令，每个Editor还有自己的undo/redo栈
 */
public class CommandManager {
    private Stack<UndoableCommand> undoStack;
    private Stack<UndoableCommand> redoStack;
    
    public CommandManager() {
        this.undoStack = new Stack<>();
        this.redoStack = new Stack<>();
    }
    
    /**
     * 执行命令
     */
    public void executeCommand(Command command) throws CommandException {
        command.execute();
        
        // 如果是可撤销命令，加入undo栈
        if (command instanceof UndoableCommand) {
            UndoableCommand undoableCommand = (UndoableCommand) command;
            if (undoableCommand.isUndoable()) {
                undoStack.push(undoableCommand);
                // 新的操作清除redo栈
                redoStack.clear();
            }
        }
    }
    
    /**
     * 撤销上一个命令
     */
    public void undo() throws CommandException {
        if (undoStack.isEmpty()) {
            throw new CommandException("没有可撤销的操作");
        }
        
        UndoableCommand command = undoStack.pop();
        command.undo();
        redoStack.push(command);
    }
    
    /**
     * 重做上一个撤销的命令
     */
    public void redo() throws CommandException {
        if (redoStack.isEmpty()) {
            throw new CommandException("没有可重做的操作");
        }
        
        UndoableCommand command = redoStack.pop();
        command.execute();
        undoStack.push(command);
    }
    
    /**
     * 检查是否可以撤销
     */
    public boolean canUndo() {
        return !undoStack.isEmpty();
    }
    
    /**
     * 检查是否可以重做
     */
    public boolean canRedo() {
        return !redoStack.isEmpty();
    }
    
    /**
     * 清空命令历史
     */
    public void clear() {
        undoStack.clear();
        redoStack.clear();
    }
}
